File: articles/john-von-neumann-and-the-architecture-of-modern-computing/go/fetch_execute.go
package main

import "fmt"

type Instruction struct {
	Op  string
	Arg int
}

func main() {
	program := []Instruction{{"LOAD", 2}, {"ADD", 3}, {"STORE", 0}, {"HALT", 0}}
	acc := 0
	for _, inst := range program {
		switch inst.Op {
		case "LOAD":
			acc = inst.Arg
		case "ADD":
			acc += inst.Arg
		case "STORE":
			fmt.Printf("store address=%d value=%d\n", inst.Arg, acc)
		case "HALT":
			fmt.Printf("halt accumulator=%d\n", acc)
			return
		}
	}
}

File: articles/john-von-neumann-and-the-architecture-of-modern-computing/typescript/fetch_execute.ts
type Instruction = { op: string; arg: number };

const program: Instruction[] = [
  { op: "LOAD", arg: 2 },
  { op: "ADD", arg: 3 },
  { op: "STORE", arg: 0 },
  { op: "HALT", arg: 0 },
];

let acc = 0;
for (const inst of program) {
  if (inst.op === "LOAD") acc = inst.arg;
  else if (inst.op === "ADD") acc += inst.arg;
  else if (inst.op === "STORE") console.log(`store address=${inst.arg} value=${acc}`);
  else if (inst.op === "HALT") {
    console.log(`halt accumulator=${acc}`);
    break;
  }
}
